File: Python/0212-word-search-ii.py
# Time:  O(m * n * l), l is height of Trie
# Space: O(l)

from collections import defaultdict

class TrieNode:
    def __init__(self):
        self.children = defaultdict(TrieNode)
        self.isWord = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        cur = self.root
        for c in word:
            cur = cur.children[c]
        cur.isWord = True

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        res, trie = [], Trie()

        node = trie.root

        for w in words:
            trie.insert(w)

        for i, row in enumerate(board):
            for j, col in enumerate(board[0]):
                self.dfs(board, node, i, j, "", res)
        return res

    def dfs(self, board, node, i, j, path, res):
        if node.isWord:
            res.append(path)
            # don't want to add this word multiple times
            node.isWord = False

        if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]):
            return

        c = board[i][j]
        if c not in node.children:
            return

        board[i][j] = '#'

        for ni, nj in (i+1, j), (i-1, j), (i, j+1), (i, j-1):
            self.dfs(board, node.children[c], ni, nj, path + c, res)

        board[i][j] = c


# Notes: In dfs the 'node' and 'path' is from previous step
# We put the cell value temporarily as # so that it won't be
# included again in the same word. Once we return we put back the value

# in most cases in dfs, cur actually refers to prev

File: C++/0212-word-search-ii.cpp
// Tags: Amazon Backtracking Trie
// TLE: Time limit exceeded (To check)
class TrieNode {
    public:
    bool isWord;
    unordered_map<char, TrieNode*> children;
    TrieNode(): isWord {false} {
        
    }
    ~TrieNode() {
        for (auto& kvp : children)
            if (kvp.second)
                delete kvp.second;
    }
};

class Trie {
    public:
    TrieNode *root;
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        auto *cur = root;
        for (auto c: word) {
            if(!cur->children.count(c))
                cur->children[c] = new TrieNode();
            cur = cur->children[c];
        }
        cur->isWord = true;
    }
};

class Solution {
public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        Trie trie;
        vector<string> res;
        for (auto w: words)
            trie.insert(w);
        int m = board.size(), n = board[0].size();
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                dfs(board, trie.root, i, j, "", res);
            }
        }
        return res;
    }
    
    void dfs(vector<vector<char>> &board, TrieNode *cur, int i, int j, string sofar, vector<string> &res) {
        if (cur->isWord) {
            res.push_back(sofar);
            // we can also use res as unordered_set
            // then we don't have to mark isWord = false
            cur->isWord = false;
            //return;
        }
        // invalid state
        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size())
            return;

        // not in trie or already visited   
        char c = board[i][j];
        if (!cur->children.count(c) || c == '#')
            return;
        
        // mark visited (temporarily)
        board[i][j] = '#';
        
        // try each direction
        vector<pair<int, int>> directions {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        for (auto &d: directions) {
            int ni = i + d.first, nj = j + d.second;
            dfs(board, cur->children[c], ni, nj, sofar + c, res);
        }

        // unmark visited
        board[i][j] = c;
    }
};
